pennylane/interfaces/jacobian_products.py

```diff
@@ -0,0 +1,173 @@
+# Copyright 2018-2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Defines classes that take the vjps, jvps, and jacobians of circuits.
+
+"""
+import abc
+from functools import partial
+from typing import Tuple, Callable, Optional
+
+import pennylane as qml
+from pennylane.tape import QuantumScript
+from pennylane.typing import ResultBatch, TensorLike
+
+Batch = Tuple[QuantumScript]
+
+
+class JacobianProductCalculator(abc.ABC):
+    """Provides methods for calculating the jvp and vjps for tapes and tangents/ cotangents."""
+
+    @abc.abstractmethod
+    def execute_and_compute_jvp(
+        self, tapes: Batch, tangents: TensorLike
+    ) -> Tuple[ResultBatch, Tuple]:
+        """Calculate both the results for a batch of tapes and the jvp.
+
+        This method is required for the jax interface.
+
+        Args:
+            tapes: The batch of tapes to take the derivatives of
+            tangents (Sequence[Sequence[TensorLike]]): the tangents for the parameters of the tape
+
+        Returns:
+            ResultBatch, TensorLike
+
+        **Examples:**
+
+        >>> tape0 = qml.tape.QuantumScript([qml.RX(0.1, wires=0)], [qml.expval(qml.PauliZ(0))])
+        >>> tape1 = qml.tape.QuantumScript([qml.RY(0.2, wires=0)], [qml.expval(qml.PauliZ(0))])
+        >>> batch = (tape0, tape1)
+        >>> tangents0 = (1.5, )
+        >>> tangents1 = (2.0, )
+        >>> tangents = (tangents0, tangents1)
+        >>> results, jvps = jp_method.execute_and_compute_jvp(batch, tangents)
+        >>> results
+        (0.9950041652780258, 0.9800665778412417)
+        >>> np.cos(0.1), np.cos(0.2)
+        (0.9950041652780258, 0.9800665778412416)
+        >>> jvps
+        (array(-0.14975012), array(-0.39733866))
+        >>> 1.5 * -np.sin(0.1), 2.0 * -np.sin(0.2)
+        (-0.14975012497024223, -0.39733866159012243)
+
+        """
+
+    @abc.abstractmethod
+    def compute_vjp(self, tapes: Batch, dy: Tuple[Tuple[TensorLike]]) -> Tuple:
+        """Compute the vjp for a given batch of tapes.
+
+        This method is used by autograd, torch, and tensorflow.
+
+        Args:
+            tapes: the batch of tapes to the the derivatives of
+            dy: the derivatives of the results of an execution
```

Comment:
Isn't the second input again a batch of tangents? :thinking:  I'm not sure I follow the current description of `dy`, aren't we computing the derivative?
```suggestion
            tapes (tuple[`~.QuantumScript`]): the batch of tapes to the the derivatives of
            dy (tuple[tuple[tensor_like]]): the derivatives of the results of an execution
```

Reply:
cotangents, not tangents. `dy` is the term most interfaces use. Jax just called it `g`, which is even less informative.

Reply:
>cotangents, not tangents

ah right, thanks.

The description sounds a bit like `dy` already knows about the derivatives of tape execution results, but it contains the derivatives of the final output with respect to the tape execution results, right?

Reply:
The derivative of the tape execution results with respect to the classical post processing... I think. https://docs.pennylane.ai/en/stable/code/api/pennylane.gradients.vjp.html

Doesn't necessarily explain it much better.